mesa/model.py

```diff
@@ -71,39 +77,95 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self.running = True
         self.schedule = None
         self.current_id = 0
-        self.agents_: defaultdict[type, dict] = defaultdict(dict)
+
+        self._setup_agent_registration()
 
         self._steps: int = 0
         self._time: TimeT = 0  # the model's clock
 
     @property
     def agents(self) -> AgentSet:
         """Provides an AgentSet of all agents in the model, combining agents from all types."""
-
-        if hasattr(self, "_agents"):
-            return self._agents
-        else:
-            all_agents = itertools.chain.from_iterable(self.agents_.values())
-            return AgentSet(all_agents, self)
+        return self._all_agents
 
     @agents.setter
     def agents(self, agents: Any) -> None:
         raise AttributeError(
-            "You are trying to set model.agents. In Mesa 3.0 and higher, this attribute will be "
+            "You are trying to set model.agents. In Mesa 3.0 and higher, this attribute is "
             "used by Mesa itself, so you cannot use it directly anymore."
             "Please adjust your code to use a different attribute name for custom agent storage."
         )
 
-        self._agents = agents
-
     @property
     def agent_types(self) -> list[type]:
         """Return a list of different agent types."""
-        return list(self.agents_.keys())
+        return list(self._agents_by_type.keys())
 
     def get_agents_of_type(self, agenttype: type[Agent]) -> AgentSet:
-        """Retrieves an AgentSet containing all agents of the specified type."""
-        return AgentSet(self.agents_[agenttype].keys(), self)
+        """Retrieves an AgentSet containing all agents of the specified type.
+
+        Args:
+            agenttype: The type of agent to retrieve.
+
+        Raises:
+            KeyError: If agenttype does not exist
+
+
+        """
+        return self._agents_by_type[agenttype]
+
+    def _setup_agent_registration(self):
+        self._agents = {}
+        self._agents_by_type: dict[type, AgentSet] = {}
+        self._all_agents = AgentSet([], self)
```

Comment:
Okay, so now we're keeping 4 records (if we count `self.agents`)? Could you explain a bit why each is needed?

I think this can be a place where some performance is lost, in models that frequently kill and create agents (like wolf-sheep)?

Reply:
self._agents contains the hard refs, so this is essential.

self._all_agents is the agentset with all agents. The main performance motivation for this PR. So again essential.

The only one you could debate is agents_by_type. This you could also do via `self._all_agents.groupby(types).groups` and thus as a property. What is better from a performance standpoint is hard to say. adding and removing stuff from dicts is relatively cheap to do.

Reply:
Thanks, appreciated. Could you add this in a comment or docstring behind/above each?

`self.agents` is the only thing we formally expose, all others are private and use-at-you-own risk.

For now I see the use case of a separate `_agents_by_type`. We can dive into the performance later.

And at least we got rid of `self.agents_`!

Reply:
> self.agents is the only thing we formally expose, all others are private and use-at-you-own risk.

Which is exactly why I moved everything else into private. It keeps the public side the same as before. 

I'll add a few comments on the datastructures later today